File: Design-2/queue_using_stacks.py

'''
Did it run on leetcode: Yes
Challenges faced: Yes

How to make enQueue and deQueue operation both 0(1). Later I realized that, I can only make
one of the operation 0(1)

Algorithm:
1 -> Take two stacks, in one you will push elements and other as temporary.
2 -> Whenever you want to add a element add into elementsWillBePushedHere stack
3 -> Whenever you want to pop, pop all the elments from elementsWillBePushedHere and push it into
temp stack. Pop the top element of temp stack, which is your front element. 
'''


class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.elementsWillBePushedHere = []
        self.tempStack = []
        

    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: None
        """
        self.elementsWillBePushedHere.append(x)
        

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        if not self.tempStack:
            
            while self.elementsWillBePushedHere:
                self.tempStack.append(self.elementsWillBePushedHere.pop())
        
        return self.tempStack.pop()
                
        
        
    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        elem = self.pop()
        self.tempStack.append(elem)
        return elem
        

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return not self.elementsWillBePushedHere and not self.tempStack
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: queueusingstacks.java
// Time Complexity : push, empty: O(1), pop, peek: Amortized O(1), worst case O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class MyQueue {

    Stack<Integer> in;
    Stack<Integer> out; //need two stacks, in stack to push elements, out stack to get first element pushed (fifo)
    
    public MyQueue() {
        in=new Stack<>();
        out=new Stack<>();
    }
    
    public void push(int x) { //push element to in stack
        in.push(x);
    }
    
    public int pop() { //
        peek();
        return out.pop();
    }
    
    public int peek() { //if out stack is empty pop from in stack to out stack then, return top element from out stack
        if(out.isEmpty()){
            while(!in.isEmpty()){
                out.push(in.pop());
            }
        }
        return out.peek();
    }
    
    public boolean empty() { //check if both stacks are empty
        return in.isEmpty() && out.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */